cli/app/api.go

```diff
@@ -442,45 +443,520 @@ func PublishToMashery(user *ApiUser, appDir string, gatewayJSON string, host str
 			fmt.Errorf("%v", err)
-			return err
+			panic(err)
 		}
+		apiId, apiName, endpoints = GetApiDetails(s)
 
-		fmt.Println("Successfully published to mashery!")
 	} else {
-		var prettyJSON bytes.Buffer
-		err := json.Indent(&prettyJSON, cleanedTfSwaggerDoc, "", "\t")
+		m := f[0].(map[string]interface{})
+		var m1 map[string]interface{}
+		json.Unmarshal(cleanedTfApiSwaggerDoc, &m1)
+		merged := merge(m, m1, 0)
+		var mergedDoc []byte
+		if mergedDoc, err = json.Marshal(merged); err != nil {
+			panic(err)
+		}
+		serviceId := merged["id"].(string)
+		s, err := user.Update("services/"+serviceId, "id,name,endpoints.id,endpoints.name,endpoints.inboundSslRequired,endpoints.outboundRequestTargetPath,endpoints.outboundTransportProtocol,endpoints.publicDomains,endpoints.requestAuthenticationType,endpoints.requestPathAlias,endpoints.requestProtocol,endpoints.supportedHttpMethods,endoints.systemDomains,endpoints.trafficManagerDomain", string(mergedDoc), token)
 		if err != nil {
-			return err
+			fmt.Fprintf(os.Stderr, "Error: Unable to update the api %s\n\n", s)
+			fmt.Errorf("%v", err)
+			panic(err)
 		}
+		apiId, apiName, endpoints = GetApiDetails(s)
 
-		fmt.Printf("%s", prettyJSON.Bytes())
-		fmt.Println("\nMocked! Did not attempt to publish.\n")
+		updated = true
 	}
 
-	return nil
+	return apiId, apiName, endpoints, updated
+}
+
+func merge(dst, src map[string]interface{}, depth int) map[string]interface{} {
+	for key, srcVal := range src {
+		if dstVal, ok := dst[key]; ok {
+			if reflect.ValueOf(dstVal).Kind() == reflect.Map {
+				srcMap, srcMapOk := mapify(srcVal)
+				dstMap, dstMapOk := mapify(dstVal)
+				if srcMapOk && dstMapOk {
+					srcVal = merge(dstMap, srcMap, depth+1)
+				}
+			} else if (key == "endpoints" || key == "plans") && reflect.ValueOf(dstVal).Kind() == reflect.Slice {
+				m_d := dstVal.([]interface{})
+				m_s := srcVal.([]interface{})
+				items := []map[string]interface{}{}
+
+				for _, d_item := range m_d {
+					i_d := d_item.(map[string]interface{})
+					var i_s map[string]interface{}
+					for _, s_item := range m_s {
+						i_s = s_item.(map[string]interface{})
+						if i_s["requestPathAlias"] == i_d["requestPathAlias"] {
+							i_s2 := merge(i_d, i_s, depth+1)
+							items = append(items, i_s2)
+						}
+					}
+				}
+
+				for _, s_item := range m_s {
+					i_s := s_item.(map[string]interface{})
+					if !MatchingEndpoint(i_s, m_d) {
+						items = append(items, i_s)
+					}
+				}
+				srcVal = items
+			}
+		}
+
+		dst[key] = srcVal
+	}
+	return dst
+}
+
+func MatchingEndpoint(ep map[string]interface{}, epList []interface{}) bool {
+	var i_d map[string]interface{}
+	for _, d_item := range epList {
+		i_d = d_item.(map[string]interface{})
+		if i_d["requestPathAlias"] == ep["requestPathAlias"] {
+			return true
+		}
+	}
+	return false
+}
+func mapify(i interface{}) (map[string]interface{}, bool) {
+	value := reflect.ValueOf(i)
+	if value.Kind() == reflect.Map {
+		m := map[string]interface{}{}
```

Comment:
Would a type assertion work here instead of reflection: https://tour.golang.org/methods/15